File: frontend/js/scope3.js
// ═══ frontend/js/scope3.js ═══

// Mapping source_type par catégorie
const SOURCE_TYPES = {
    1: [
        { value: "steel_raw", label: "Acier brut", unit: "kg", factor: 1.85 },
        { value: "aluminum_raw", label: "Aluminium brut", unit: "kg", factor: 8.24 },
        { value: "cement_clinker", label: "Clinker ciment", unit: "kg", factor: 0.83 },
        { value: "iron_ore", label: "Minerai de fer", unit: "kg", factor: 0.02 },
        { value: "coal_coke", label: "Coke de charbon", unit: "kg", factor: 3.09 },
        { value: "limestone", label: "Calcaire", unit: "kg", factor: 0.012 },
        { value: "ammonia", label: "Ammoniac", unit: "kg", factor: 2.16 },
        { value: "bauxite", label: "Bauxite", unit: "kg", factor: 0.007 },
        { value: "scrap_metal", label: "Ferraille recyclée", unit: "kg", factor: 0.42 },
    ],
    3: [
        { value: "electricity_upstream", label: "Pertes réseau électrique", unit: "kWh", factor: 0.108 },
        { value: "fuel_upstream", label: "Extraction/raffinage fuel", unit: "litres", factor: 0.59 },
        { value: "gas_upstream", label: "Extraction/transport gaz", unit: "m3", factor: 0.41 },
    ],
    4: [
        { value: "transport_truck", label: "Camion", unit: "t.km", factor: 0.096 },
        { value: "transport_rail", label: "Train fret", unit: "t.km", factor: 0.028 },
        { value: "transport_ship", label: "Cargo maritime", unit: "t.km", factor: 0.016 },
        { value: "transport_air", label: "Fret aérien", unit: "t.km", factor: 0.602 },
    ],
    5: [
        { value: "waste_landfill", label: "Mise en décharge", unit: "kg", factor: 0.586 },
        { value: "waste_incineration", label: "Incinération", unit: "kg", factor: 0.021 },
        { value: "waste_recycling", label: "Recyclage", unit: "kg", factor: 0.010 },
    ],
    9: [
        { value: "transport_truck_out", label: "Camion (sortant)", unit: "t.km", factor: 0.096 },
        { value: "transport_ship_out", label: "Maritime (sortant)", unit: "t.km", factor: 0.016 },
        { value: "transport_rail_out", label: "Train (sortant)", unit: "t.km", factor: 0.028 },
    ],
    12: [
        { value: "end_of_life_steel", label: "Recyclage acier", unit: "kg", factor: 0.15 },
        { value: "end_of_life_aluminum", label: "Recyclage aluminium", unit: "kg", factor: 0.30 },
        { value: "end_of_life_cement", label: "Démolition béton", unit: "kg", factor: 0.01 },
        { value: "end_of_life_landfill", label: "Mise en décharge", unit: "kg", factor: 0.50 },
    ],
};

function updateFormFields() {
    const category = parseInt(document.getElementById('scope3-category').value);
    const sourceSelect = document.getElementById('scope3-source-type');
    
    // Remplir les types de source
    sourceSelect.innerHTML = '';
    const sources = SOURCE_TYPES[category] || [];
    sources.forEach(s => {
        const opt = document.createElement('option');
        opt.value = s.value;
        opt.textContent = `${s.label} (${s.factor} kg CO₂/${s.unit})`;
        sourceSelect.appendChild(opt);
    });
    
    // Afficher/masquer champs transport
    const transportFields = document.getElementById('transport-fields');
    transportFields.style.display = [4, 9].includes(category) ? 'block' : 'none';
    
    // Afficher/masquer champs fournisseur
    const supplierFields = document.getElementById('supplier-fields');
    supplierFields.style.display = category === 1 ? 'block' : 'none';
    
    // Mettre à jour l'unité
    if (sources.length > 0) {
        document.getElementById('scope3-unit').value = sources[0].unit;
    }
}

async function submitScope3Entry() {
    const category = parseInt(document.getElementById('scope3-category').value);
    
    const payload = {
        category: category,
        direction: category <= 8 ? "upstream" : "downstream",
        description: document.getElementById('scope3-description').value,
        source_type: document.getElementById('scope3-source-type').value,
        quantity: parseFloat(document.getElementById('scope3-quantity').value),
        unit: document.getElementById('scope3-unit').value,
        date: document.getElementById('scope3-date').value,
        data_quality: document.getElementById('scope3-data-quality').value,
        source_document: document.getElementById('scope3-source-doc').value || null,
    };
    
    // Champs transport
    if ([4, 9].includes(category)) {
        payload.distance_km = parseFloat(document.getElementById('scope3-distance').value) || null;
        payload.weight_tonnes = parseFloat(document.getElementById('scope3-weight').value) || null;
        payload.transport_mode = document.getElementById('scope3-transport-mode').value;
        payload.origin = document.getElementById('scope3-origin').value || null;
        payload.destination = document.getElementById('scope3-destination').value || null;
    }
    
    // Champs fournisseur
    if (category === 1) {
        payload.supplier_name = document.getElementById('scope3-supplier-name').value || null;
        payload.supplier_country = document.getElementById('scope3-supplier-country').value || null;
    }
    
    try {
        const response = await fetch('http://localhost:8000/scope3/entries', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify(payload)
        });
        
        const result = await response.json();
        
        const resultBox = document.getElementById('scope3-result');
        resultBox.style.display = 'block';
        resultBox.innerHTML = `
            <div class="result-success">
                <strong>✅ ${result.message}</strong><br>
                Catégorie : ${result.category_name}<br>
                CO₂ : ${result.co2_kg} kg (${result.co2_tonnes} tonnes)<br>
                Facteur : ${result.emission_factor} kg CO₂/unité
            </div>
        `;
        
        // Rafraîchir le dashboard
        loadScope3Data();
        
    } catch (error) {
        console.error('Erreur:', error);
        document.getElementById('scope3-result').innerHTML = 
            `<div class="result-error">❌ Erreur: ${error.message}</div>`;
    }
}

function switchDirection(direction) {
    document.querySelectorAll('.tab').forEach(t => t.classList.remove('active'));
    document.querySelector(`[data-direction="${direction}"]`).classList.add('active');
    
    const select = document.getElementById('scope3-category');
    // Auto-sélectionner la première option de la direction
    if (direction === 'upstream') select.value = '1';
    else select.value = '9';
    
    updateFormFields();
}

// Initialisation
document.addEventListener('DOMContentLoaded', () => {
    updateFormFields();
});

File: routers/scope3.py
# ══════════════════════════════════════════
# routers/scope3.py — CRUD Scope 3
# ══════════════════════════════════════════

from fastapi import APIRouter, HTTPException, Query
from database import get_connection as get_db_connection
from datetime import date, datetime
from typing import Optional

router = APIRouter(prefix="/scope3", tags=["Scope 3"])

# ── Mapping facteurs (sera remplacé par DB lookup) ──
SCOPE3_FACTORS = {
    # Cat 1 - Matières premières
    "steel_raw": 1.85, "aluminum_raw": 8.24, "cement_clinker": 0.83,
    "iron_ore": 0.02, "coal_coke": 3.09, "limestone": 0.012,
    "ammonia": 2.16, "bauxite": 0.007, "scrap_metal": 0.42,
    # Cat 3 - Énergie amont
    "electricity_upstream": 0.108, "fuel_upstream": 0.59, "gas_upstream": 0.41,
    # Cat 4 & 9 - Transport
    "transport_truck": 0.096, "transport_rail": 0.028,
    "transport_ship": 0.016, "transport_air": 0.602,
    "transport_truck_out": 0.096, "transport_ship_out": 0.016,
    "transport_rail_out": 0.028,
    # Cat 5 - Déchets
    "waste_landfill": 0.586, "waste_incineration": 0.021, "waste_recycling": 0.010,
    # Cat 12 - Fin de vie
    "end_of_life_steel": 0.15, "end_of_life_aluminum": 0.30,
    "end_of_life_cement": 0.01, "end_of_life_landfill": 0.50,
}

SCOPE3_CATEGORIES = {
    1: "Matières premières achetées",
    3: "Énergie amont (hors S1/S2)",
    4: "Transport entrant",
    5: "Déchets de production",
    9: "Transport sortant",
    12: "Fin de vie produits vendus",
}


# ── POST /scope3/entries — Ajouter une entrée ──
@router.post("/entries")
async def create_scope3_entry(entry: dict):
    """Ajouter une entrée Scope 3 avec calcul automatique CO2"""
    
    source_type = entry.get("source_type")
    quantity = entry.get("quantity", 0)
    category = entry.get("category")
    
    # Validation catégorie
    if category not in SCOPE3_CATEGORIES:
        raise HTTPException(400, f"Catégorie {category} non supportée. Valides: {list(SCOPE3_CATEGORIES.keys())}")
    
    # Récupérer facteur
    factor = SCOPE3_FACTORS.get(source_type)
    if not factor:
        raise HTTPException(400, f"Type source '{source_type}' inconnu. Valides: {list(SCOPE3_FACTORS.keys())}")
    
    # Calcul CO2
    # Pour transport : quantity = t.km (poids × distance)
    if category in [4, 9] and entry.get("distance_km") and entry.get("weight_tonnes"):
        effective_quantity = entry["distance_km"] * entry["weight_tonnes"]
    else:
        effective_quantity = quantity
    
    co2_kg = effective_quantity * factor
    
    # Direction automatique
    direction = "upstream" if category <= 8 else "downstream"
    
    conn = get_db_connection()
    cur = conn.cursor()
    
    cur.execute("""
        INSERT INTO scope3_entries 
        (category, category_name, direction, description, source_type,
         quantity, unit, emission_factor, co2_kg,
         distance_km, weight_tonnes, transport_mode, origin, destination,
         supplier_name, supplier_country, date, data_quality, source_document,
         period_quarter, period_year)
        VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
        RETURNING id
    """, (
        category, SCOPE3_CATEGORIES[category], direction,
        entry.get("description", ""),
        source_type, effective_quantity, entry.get("unit", ""),
        factor, co2_kg,
        entry.get("distance_km"), entry.get("weight_tonnes"),
        entry.get("transport_mode"), entry.get("origin"), entry.get("destination"),
        entry.get("supplier_name"), entry.get("supplier_country"),
        entry.get("date"), entry.get("data_quality", "estimated"),
        entry.get("source_document"),
        entry.get("period_quarter"), entry.get("period_year")
    ))
    
    new_id = cur.fetchone()["id"]
    conn.commit()
    cur.close()
    conn.close()
    
    return {
        "id": new_id,
        "co2_kg": round(co2_kg, 2),
        "co2_tonnes": round(co2_kg / 1000, 4),
        "emission_factor": factor,
        "category": category,
        "category_name": SCOPE3_CATEGORIES[category],
        "direction": direction,
        "message": f"✅ Entrée Scope 3 créée — {round(co2_kg, 2)} kg CO2"
    }


# ── GET /scope3/entries — Lister les entrées ──
@router.get("/entries")
async def list_scope3_entries(
    direction: Optional[str] = None,
    category: Optional[int] = None
):
    """Liste des entrées Scope 3 actives"""
    conn = get_db_connection()
    cur = conn.cursor()
    
    query = "SELECT * FROM scope3_entries WHERE actif = true"
    params = []
    
    if direction:
        query += " AND direction = %s"
        params.append(direction)
    if category:
        query += " AND category = %s"
        params.append(category)
    
    query += " ORDER BY date DESC"
    cur.execute(query, params)
    
    rows = [dict(row) for row in cur.fetchall()]
    
    cur.close()
    conn.close()
    return rows


# ── GET /scope3/summary — Résumé Scope 3 ──
@router.get("/summary")
async def scope3_summary():
    """Résumé complet des émissions Scope 3"""
    conn = get_db_connection()
    cur = conn.cursor()
    
    # Total par direction
    cur.execute("""
        SELECT direction, SUM(co2_kg) as total
        FROM scope3_entries WHERE actif = true
        GROUP BY direction
    """)
    direction_totals = {row["direction"]: float(row["total"]) for row in cur.fetchall()}

    # Total par catégorie
    cur.execute("""
        SELECT category, SUM(co2_kg) AS total
        FROM scope3_entries WHERE actif = true
        GROUP BY category ORDER BY category
    """)
    category_totals = {str(row["category"]): round(float(row["total"]), 2) for row in cur.fetchall()}

    # Qualité des données
    cur.execute("""
        SELECT data_quality, COUNT(*) AS cnt
        FROM scope3_entries WHERE actif = true
        GROUP BY data_quality
    """)
    quality = {row["data_quality"]: int(row["cnt"]) for row in cur.fetchall()}

    # Count
    cur.execute("SELECT COUNT(*) AS cnt FROM scope3_entries WHERE actif = true")
    count = cur.fetchone()["cnt"]
    
    upstream = direction_totals.get("upstream", 0) or 0
    downstream = direction_totals.get("downstream", 0) or 0
    total = upstream + downstream
    
    cur.close()
    conn.close()
    
    return {
        "total_co2_kg": round(total, 2),
        "total_co2_tonnes": round(total / 1000, 4),
        "upstream_co2_kg": round(upstream, 2),
        "upstream_co2_tonnes": round(upstream / 1000, 4),
        "downstream_co2_kg": round(downstream, 2),
        "downstream_co2_tonnes": round(downstream / 1000, 4),
        "by_category": category_totals,
        "entry_count": count,
        "data_quality_breakdown": quality
    }


# ── GET /scope3/factors — Facteurs disponibles ──
@router.get("/factors")
async def scope3_factors():
    """Liste tous les facteurs d'émission Scope 3 disponibles"""
    result = {}
    for cat_id, cat_name in SCOPE3_CATEGORIES.items():
        result[cat_id] = {
            "name": cat_name,
            "factors": {
                k: {"factor": v, "unit": "kg CO2/unité"}
                for k, v in SCOPE3_FACTORS.items()
            }
        }
    return result


# ── GET /scope3/categories — Catégories actives ──
@router.get("/categories")
async def scope3_categories():
    """Liste les catégories Scope 3 actives dans CarbonIQ"""
    return {
        "upstream": [
            {"id": 1, "name": "Matières premières achetées", "ghg": "Purchased goods & services"},
            {"id": 3, "name": "Énergie amont (hors S1/S2)", "ghg": "Fuel & energy-related"},
            {"id": 4, "name": "Transport entrant", "ghg": "Upstream transportation"},
            {"id": 5, "name": "Déchets de production", "ghg": "Waste generated in operations"},
        ],
        "downstream": [
            {"id": 9, "name": "Transport sortant", "ghg": "Downstream transportation"},
            {"id": 12, "name": "Fin de vie produits", "ghg": "End-of-life treatment"},
        ]
    }


# ── DELETE (soft) /scope3/entries/{id} ──
@router.post("/entries/{entry_id}/supprimer")
async def soft_delete_scope3(entry_id: int, body: dict):
    """Soft delete Scope 3 (ISO 14064 compliant)"""
    raison = body.get("raison")
    if not raison:
        raise HTTPException(400, "Raison obligatoire pour la traçabilité CBAM")
    
    conn = get_db_connection()
    cur = conn.cursor()
    
    cur.execute("""
        UPDATE scope3_entries SET actif = false, raison = %s, updated_at = NOW()
        WHERE id = %s
    """, (raison, entry_id))
    
    # Journal
    cur.execute("""
        INSERT INTO journal_scope3 (scope3_entry_id, champ_modifie, ancienne_val, nouvelle_val, raison)
        VALUES (%s, 'actif', 'true', 'false', %s)
    """, (entry_id, raison))
    
    conn.commit()
    cur.close()
    conn.close()
    
    return {"message": f"Entrée {entry_id} désactivée", "raison": raison}